### PR TITLE
SDK-6020: C# plugin

### PR DESCRIPTION
**Pull request by liviuivan-neurala**
**JIRA**: https://neurala.atlassian.net/browse/SDK-6020

Changelog
------------------

- adapt existing websocket plugin for production use
- add C# server library & executable

Reviewer Checklist
------------------

> Should be edited by reviewers only

- [ ] Does this implement the story reqs?
- [ ] Does this match the [coding standard](../wiki/C-Based-Languages-Style-Guide)?
	- [ ] Are all TODO's in the [proper format](../wiki/C-Based-Languages-Style-Guide#commenting)?
	- [ ] Have JIRA tickets been made for all TODO's?
- [ ] Does this include tests that exercise the altered/added code?
- [ ] If it modifies a public API, is the documentation updated?

### CODE REVIEW

File: plugins/websocket/include/websocket/Client.h [hunk 1]
```diff
@@ -32,29 +33,25 @@
...
 	Client(Client&&) = default;
 	Client& operator=(const Client&) = delete;
 	Client& operator=(Client&&) = delete; // the stream is not move-assignable
 
 	/// Closes the WebSocket stream used to communicate with the server.
-	~Client() { m_stream.close(beast::websocket::close_code::normal); }
+	~Client()
+	{
+		boost::system::error_code ec;
+		m_stream.close(boost::beast::websocket::close_code::normal, ec);
```
Comment (jraynal) [inline plugins/websocket/include/websocket/Client.h:53]: We may want to log that `ec`?

File: plugins/websocket/servers/csharp/Binaries/.gitattributes [hunk 2]
```diff
@@ -0,0 +1,4 @@
+Neurala.VIA.WebSocket.dll filter=lfs diff=lfs merge=lfs -text
+Newtonsoft.Json.dll filter=lfs diff=lfs merge=lfs -text
+Test.exe filter=lfs diff=lfs merge=lfs -text
+websocket-sharp.dll filter=lfs diff=lfs merge=lfs -text
```
Comment (jraynal) [inline plugins/websocket/servers/csharp/Binaries/.gitattributes:4]: @Deneuki can we more that to top level, and mark any dll? Should be something like: `*.dll filter=lfs diff=lfs merge=lfs -text`

File: plugins/websocket/servers/csharp/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+
+add_library(Neurala.VIA.WebSocket Sources/Server.cs)
+
+set_target_properties(Neurala.VIA.WebSocket PROPERTIES 
+    VS_DOTNET_REFERENCE_Newtonsoft.Json
+        "Binaries/Newtonsoft.Json.dll"
+    VS_DOTNET_REFERENCE_websocket-sharp
+        "Binaries/websocket-sharp.dll")
+
+add_executable(Neurala.VIA.Test Sources/Test.cs)
```
Comment (jraynal) [inline plugins/websocket/servers/csharp/CMakeLists.txt:13]: @Deneuki can delete.

File: plugins/websocket/servers/csharp/Makefile [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+SOURCES := Sources
+
+test : $(BINARIES)/Test.exe
+websocket : $(BINARIES)/Neurala.VIA.WebSocket.dll
+
+$(BINARIES)/Test.exe : $(BINARIES)/Neurala.VIA.WebSocket.dll $(SOURCES)/Test.cs
+	$(CSC) $(CSCFLAGS) /out:$@ /t:exe /r:$(BINARIES)/Neurala.VIA.WebSocket.dll $(filter %.cs,$^)
+
+$(BINARIES)/Neurala.VIA.WebSocket.dll : $(SOURCES)/Server.cs
+	$(CSC) $(CSCFLAGS) /out:$@ /t:library /r:$(BINARIES)/Newtonsoft.Json.dll /r:$(BINARIES)/websocket-sharp.dll $(filter %.cs,$^)
```
Comment (jraynal) [inline plugins/websocket/servers/csharp/Makefile:14]: @Deneuki can delete too.

**Review by jraynal — CHANGES_REQUESTED** [review]
@Deneuki do we need all those files in `**/csharp/*/{obj,bin,...}`?

File: plugins/websocket/include/websocket/Client.h (hunk 1 above)
Comment (liviuivan-neurala) [reply plugins/websocket/include/websocket/Client.h]: Don't see why not. Done.

**Comment by liviuivan-neurala** [comment]
Moved previous state of this PR to a new branch (C#_JPG_plugin - anyone feel free to rename it) and kept the more narrow improvements (exception safety, single server, etc.).

File: plugins/websocket/src/InitMe.cpp [hunk 5]
```diff
@@ -23,26 +23,26 @@
...
 		return nullptr;
 	}
-	*status = pm.registerPlugin<plug::ws::Input>("Input", Version(1, 0));
+	*status = pm.registerPlugin<plug::ws::Input>("websocketInput", Version(1, 0));
 	if (*status != PluginStatus::success())
 	{
 		return nullptr;
 	}
-	*status = pm.registerPlugin<plug::ws::Output>("Output", Version(1, 0));
+	*status = pm.registerPlugin<plug::ws::Output>("websocketOutput", Version(1, 0));
```
Comment (jraynal) [inline plugins/websocket/src/InitMe.cpp:45]: Why is the first one `libwebsocket`, and the other are only `websocket`?

Comment (liviuivan-neurala) [reply plugins/websocket/src/InitMe.cpp:45]: Looking at `src\video\VideoPluginDiscoverer.cpp` now, I don't see the reason, though this is how I got it working at the time the changes were made.

I couldn't get the Inspector to work properly for testing yesterday, I am trying again today. If your environment is set up for it, could you please verify the plugin as well?

Comment (liviuivan-neurala) [reply plugins/websocket/src/InitMe.cpp]: Fixed the naming (now it works fine) and changed the socket number (for whatever reason the old one crashes the acceptor on my PC). After that I was able to test it; it is ready to merge in my opinion.

**Review by jraynal — APPROVED** [review]

**Pull request merged**